File: Widget/app/src/main/java/com/example/lab/MainActivity.java
package com.example.lab;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;

import com.example.lab.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String num1 = binding.editText1.getText().toString();
                String num2 = binding.editText2.getText().toString();
                int sum = Integer.parseInt(num1) + Integer.parseInt(num2);
                int sub = Integer.parseInt(num1) - Integer.parseInt(num2);
                int mul = Integer.parseInt(num1) * Integer.parseInt(num2);
                double div = Double.parseDouble(num1) / Double.parseDouble(num2);

                binding.textView.setText(String.format("%d, %d %d, %.1f", sum, sub, mul, div));
            }
        });
    }
}

File: FirebaseCloudMessaging/app/src/main/java/com/example/cloudmessaging/MainActivity.kt
package com.example.cloudmessaging

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import com.example.cloudmessaging.databinding.ActivityMainBinding
import com.google.firebase.messaging.FirebaseMessaging

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        FirebaseMessaging.getInstance().token.addOnCompleteListener {
            binding.fcmToken.text = if (it.isSuccessful) it.result else "Token Error!"

            // copy FCM token to clipboard
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("FCM Token", binding.fcmToken.text)
            clipboard.setPrimaryClip(clip)

            // write to logcat
            Log.d(MyFirebaseMessagingService.TAG, "FCM token: ${binding.fcmToken.text}")
        }

        val msgBody = intent.getCharSequenceExtra("message")
        if (msgBody != null)
            binding.msgBody.text = msgBody
    }
}